### PR TITLE
[lifecycle] implemented lifecycles, added possibilities of sideeffects

### PR DESCRIPTION
**Pull request by plusgut**

### CODE REVIEW

File: src/instances/types/Array/reconcile.ts [hunk 1]
```diff
@@ -32,7 +32,7 @@ export default function (newAbstractElements: PlusnewAbstractElement[], instance
       if (!found) {
```
Comment (Reloaded2) [inline src/instances/types/Array/reconcile.ts]: für bessere Lesbarkeit bitte hier "(found === false)" verwenden .

Comment (plusgut) [reply src/instances/types/Array/reconcile.ts]: > für bessere Lesbarkeit bitte hier "(found === false)" verwenden .
agreed

please switch to english in this public repository

File: test/integration/render/lifecycle.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+      expect(container.childNodes.length).toBe(1);
+      expect((container.childNodes[0] as HTMLElement).tagName).toBe('DIV');
+      expect(componentDidMountSpy.calls.count()).toBe(1);
+    });
+
+    it('hook doing sideeffect to parent', () => {
+      const renderSpy = jasmine.createSpy('renderSpy', (Props: Props<{value: number}>, componentInstance) => {
+        componentInstance.registerLifecycleHook('componentDidMount', () => {
+          debugger;
```
Comment (gaggala) [inline test/integration/render/lifecycle.spec.tsx]: please delete debugger

File: test/integration/render/lifecycle.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+      expect((container.childNodes[0] as HTMLElement).innerHTML).toBe('1');
+      expect((container.childNodes[1] as HTMLElement).tagName).toBe('SPAN');
+      expect((container.childNodes[1] as HTMLElement).innerHTML).toBe('1');
+    });
+  });
+
+  it('hook doing sideeffect to parent', () => {
+    const renderSpy = jasmine.createSpy('renderSpy', (Props: Props<{value: number}>, componentInstance) => {
+      componentInstance.registerLifecycleHook('componentDidMount', () => {
+        debugger;
```
Comment (gaggala) [inline test/integration/render/lifecycle.spec.tsx]: please delete debugger

File: test/integration/render/lifecycle.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+      expect(componentDidMountSpy.calls.count()).toBe(0);
+
+      local.dispatch(true);
+
+      expect(container.childNodes.length).toBe(1);
+      expect((container.childNodes[0] as HTMLElement).tagName).toBe('DIV');
+      expect(componentDidMountSpy.calls.count()).toBe(1);
+    });
+
+    it('hook doing sideeffect to parent', () => {
```
Comment (gaggala) [inline test/integration/render/lifecycle.spec.tsx]: more meaningful? "First nested component dispatches a value to local store in parent and gets rerendered and displayed in DOM only once."

File: test/integration/render/lifecycle.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+      expect(renderSpy.calls.count()).toBe(1);
+      expect(anotherRenderSpy.calls.count()).toBe(1);
+      expect((container.childNodes[0] as HTMLElement).tagName).toBe('DIV');
+      expect((container.childNodes[0] as HTMLElement).innerHTML).toBe('1');
+      expect((container.childNodes[1] as HTMLElement).tagName).toBe('SPAN');
+      expect((container.childNodes[1] as HTMLElement).innerHTML).toBe('1');
+    });
+  });
+
+  it('hook doing sideeffect to parent', () => {
```
Comment (gaggala) [inline test/integration/render/lifecycle.spec.tsx]: more meaningful? "Second nested component dispatches a value to local store in parent and gets rerendered and displayed in DOM only once."

**Pull request merged**